### PR TITLE
Backup and restore callback

### PR DESCRIPTION
**Pull request by kamathis4**
Based on the [callback issue](https://github.com/keras-team/keras-core/issues/23)

### CODE REVIEW

File: keras_core/callbacks/backup_and_restore_callback.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+                f"Unrecognized save_freq: {self.save_freq}. "
+                "Expected save_freq are 'epoch' or integer values"
+            )
+
+    def on_train_begin(self, logs=None):
+        """
+        Get training state from temporary file and restore it
+        """
+        super().on_train_begin()
+        self.set_model(keras_core.saving.load_model(filepath=self.backup_dir))
```
Comment (fchollet) [inline keras_core/callbacks/backup_and_restore_callback.py:88]: Is this the right approach? Should we rather take `self.model` and load its state?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras_core/callbacks/backup_and_restore_callback.py (hunk 1 above)
Comment (kamathis4) [reply keras_core/callbacks/backup_and_restore_callback.py]: Not able to directly use **self.model** but **self._model** seems to be working

File: keras_core/callbacks/backup_and_restore_callback.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+                f"Unrecognized save_freq: {self.save_freq}. "
+                "Expected save_freq are 'epoch' or integer values"
+            )
+
+    def on_train_begin(self, logs=None):
+        """
+        Get training state from temporary file and restore it
+        """
+        if self._check_checkpoints_exists(self.file_path):
+            self._model = keras_core.saving.load_model(filepath=self.file_path)
```
Comment (sampathweb) [inline keras_core/callbacks/backup_and_restore_callback.py:113]: This will create a new instance of `self._model`.  This will not change the actual model used for training. What we want is to load weights to existing model with something like `self._model.load_weights(...)`

File: keras_core/callbacks/backup_and_restore_callback.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        logs = logs or {}
+
+        filepath = self._get_file_path(epoch, batch, logs)
+        # Create host directory if it doesn't exist.
+        dirname = os.path.dirname(filepath)
+        if dirname and not file_utils.exists(dirname):
+            file_utils.makedirs(dirname)
+
+        try:
+            self._model.save(filepath, overwrite=True)
```
Comment (sampathweb) [inline keras_core/callbacks/backup_and_restore_callback.py:152]: if we plan to use `load_weights` we may also want to use `save_weights`

File: keras_core/callbacks/backup_and_restore_callback.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        save_before_preemption: A boolean value instructing whether to turn on
+          the automatic checkpoint saving for preemption/maintenance events.
+    """
+
+    def __init__(
+        self,
+        file_path,
+        save_freq="epoch",
+        delete_checkpoint=True,
+        save_before_preemption=False,
```
Comment (sampathweb) [inline keras_core/callbacks/backup_and_restore_callback.py:83]: Looks like this arg is not used.  If we are not porting logic for `save_before_preemption` we should remove it. I am OK to remove it for now and add it back later if needed.  If we want to keep the backward compatibility on args, we should add a warning if it's set to True that its deprecated and not used.

**Review by sampathweb — CHANGES_REQUESTED** [review]

**Comment by sampathweb** [comment]
Thanks for making the changes.  I will review it today.

File: keras_core/callbacks/backup_and_restore_callback.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+    def on_train_begin(self, logs=None):
+        """
+        Get training state from temporary file and restore it
+        """
+        if self._check_checkpoints_exists(self.file_path):
+            self._model = keras_core.saving.load_model(filepath=self.file_path)
+
+    def on_train_end(self, logs=None):
+        if self.delete_checkpoint and \
```
[the remark this answers is not in the document]
Comment (kamathis4) [reply keras_core/callbacks/backup_and_restore_callback.py]: I think this is from an old file, don't see it in my code 😅

File: keras_core/callbacks/backup_and_restore_callback.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+            self._model.load_weights(filepath=self.file_path)
+
+    def on_train_end(self, logs=None):
+        if self.delete_checkpoint and self._check_checkpoints_exists(
+            self.file_path
+        ):
+            self._cleanup_checkpoint()
+
+    def on_epoch_begin(self, epoch, logs=None):
+        if self.delete_checkpoint and self._check_checkpoints_exists(
```
Comment (sampathweb) [inline keras_core/callbacks/backup_and_restore_callback.py:124]: Why do we delete checkpoint on Epoch begin?  If an interrupt happens during training, it will not be able to recover.  Deletion of checkpoint should only happen in `on_train_end`. I checked to confirm that this logic of deleting in `on_epoch_begin` is not there in `tf.keras` implementation.

**Review by sampathweb — CHANGES_REQUESTED** [review]

File: keras_core/callbacks/backup_and_restore_callback.py (hunk 6 above)
Comment (kamathis4) [reply keras_core/callbacks/backup_and_restore_callback.py:124]: The issue I was trying to consider is that there should be only a single checkpoint at a time according to the [documentation](https://keras.io/api/callbacks/backup_and_restore/), but I understand your point, if the saving frequency is epoch and the model get interrupted between the epoch it will not be able to recover.
So I'm thinking of adding it in [here](https://github.com/keras-team/keras-core/pull/701/files#diff-c3104e6325115d3d66065a21adcafa59c58673e1412b678e2f17b1a5dfedd869R156) i.e. just before saving the weights.
Then also delete it at train end.

Comment (sampathweb) [reply keras_core/callbacks/backup_and_restore_callback.py:124]: With `overwrite=True`, why do we need to delete before writing again?

Comment (sampathweb) [reply keras_core/callbacks/backup_and_restore_callback.py:124]: After removing the delete checkpoint logic in `on_epoch_begin`, I am good with this PR.

**Review by sampathweb — CHANGES_REQUESTED** [review]

File: keras_core/callbacks/backup_and_restore_callback.py (hunk 6 above)
Comment (kamathis4) [reply keras_core/callbacks/backup_and_restore_callback.py]: done 🥸

File: keras_core/callbacks/backup_and_restore_callback.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    ...                     epochs=10, batch_size=1, callbacks=[callback],
+    ...                     verbose=0)
+    >>> # Only 6 more epochs are run, since first training got interrupted at
+    >>> # zero-indexed epoch 4, second training will continue from 4 to 9.
+    >>> len(history.history['loss'])
+    >>> 6
+
+    Args:
+        file_path: String, path to store the checkpoint.
+          e.g. `backup_dir = os.path.join(working_dir, "backup")`.
```
Comment (fchollet) [inline keras_core/callbacks/backup_and_restore_callback.py:57]: - Please use 4 space indent
- We should stick to the original argument, `backup_dir`. Then we can use it to generate the filepath internally. Note that the filepath needs to end in `.weights.h5`.

File: keras_core/callbacks/backup_and_restore_callback.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    arguments passed to compile or fit, the checkpoint saved for fault tolerance
+    can become invalid.
+
+    Example:
+
+    >>> class InterruptingCallback(keras.callbacks.Callback):
+    ...   def on_epoch_begin(self, epoch, logs=None):
+    ...     if epoch == 4:
+    ...       raise RuntimeError('Interrupting!')
+    >>> callback = keras.callbacks.BackupAndRestore(backup_dir="/tmp/backup")
```
Comment (fchollet) [inline keras_core/callbacks/backup_and_restore_callback.py:38]: (This is what is being shown in this code example)

File: keras_core/callbacks/backup_and_restore_callback.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+                    f"directory: f{filepath}"
+                )
+            # Re-throw the error for any other causes.
+            raise e
+
+    def _get_file_path(self, epoch, batch, logs):
+        """Returns the file path for checkpoint."""
+
+        try:
+            # `filepath` may contain placeholders such as
```
Comment (fchollet) [inline keras_core/callbacks/backup_and_restore_callback.py:175]: Since the filename will be autogenerated you can just include `epoch` in the filename you pick -- e.g. `backup_at_epoch_{epoch}.weights.h5`

File: keras_core/callbacks/backup_and_restore_callback.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+            )
+        if self.save_before_preemption:
+            warnings.warn("`save_before_preemption` not yet implemented")
+
+    def on_train_begin(self, logs=None):
+        """
+        Get training state from temporary file and restore it
+        """
+        if self._check_checkpoints_exists(self.file_path):
+            self._model.load_weights(filepath=self.file_path)
```
Comment (sampathweb) [inline keras_core/callbacks/backup_and_restore_callback.py:115]: We are not restoring the "Current Epoch Number", so the `model.fit` that happens after the restore re-trains for all the epochs.  In `tf.keras` implementation, it restores the current epoch and `model.fit` only trains for the remaining epochs.  See example in https://keras.io/api/callbacks/backup_and_restore/

File: keras_core/callbacks/backup_and_restore_callback_test.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+        except RuntimeError:
+            self.assertTrue(os.path.exists(filepath))
+            self.assertEqual(cbk._current_epoch, 0)
+            self.assertEqual(cbk._last_batch_seen, 1)
+
+            hist = model.fit(
+                x_train, y_train, batch_size=4, callbacks=[cbk], epochs=5
+            )
+
+            self.assertEqual(cbk._current_epoch, 4)
```
Comment (sampathweb) [inline keras_core/callbacks/backup_and_restore_callback_test.py:122]: Can you also assert that `len(hist["loss"]) == 3` to confirm that it only ran for 3 additional epochs and not the entire 5 epochs again?  I would expect that assertion to fail since we don't seem to restore the `epoch` in `on_train_begin`.

Comment (sampathweb) [reply keras_core/callbacks/backup_and_restore_callback_test.py:122]: @fchollet - What do you think we should do?  Is there a good way we can set the `iniital_epoch` of the model?  Also, we are not restoring the Optimizer state that `tf.keras` had restored when it saved it in training checkpoint.  Any suggestions.

**Review by sampathweb — CHANGES_REQUESTED** [review]

File: keras_core/callbacks/backup_and_restore_callback_test.py (hunk 11 above)
Comment (sampathweb) [reply keras_core/callbacks/backup_and_restore_callback_test.py:122]: Checked with Francois and addressed both questions - 
1. Model saving also includes optimizer, so optimizer state is also restored when backup is restored.
2. On setting `initial_epoch`, we may need to add that hook on `model` before we can set it in the `on_train_begin` of `BackupAndRestore`.  We will address that in a subsequent PR.
We can make this conversation resolved.

File: keras_core/callbacks/backup_and_restore_callback.py (hunk 10 above)
Comment (sampathweb) [reply keras_core/callbacks/backup_and_restore_callback.py:115]: Marking it resolved here.  We will address it in subsequent PR when we add the hook in `model` to set `initial_epoch`.

File: keras_core/callbacks/backup_and_restore_callback.py (hunk 7 above)
Comment (sampathweb) [reply keras_core/callbacks/backup_and_restore_callback.py:57]: @kamathis4 - Can you address this comment

**Comment by fchollet** [comment]
Hi @kamathis4 -- are you still working on this PR? Should we take it over?

**Comment by kamathis4** [comment]
> Hi @kamathis4 -- are you still working on this PR? Should we take it over?

Hey @fchollet, please do take over. Have been really busy with work and don't seem to be able to make time for oss.

**Comment by fchollet** [comment]
Ok -- merging and patching up on top. Thanks for the contribution!
